### ISSUE TITLE
Comment on IG tutorial on text

### ISSUE THREAD

**Issue by heytitle** [pytorch/captum#407]
According to [a recommendation from the authors of Integrated Gradients][1], it is written that we should use the softmax output for Integrated Gradients. 

> or multi-class classification models, the prediction head is typically a softmax operator on a 'logits' tensor. The attribution must be computed from this softmax output and not the 'logits' tensor. — See [[1]]'s [Identifying the output tensor][1].

However, on [IMDB TorchText Interpret][2], we use the logit value, i.e.  we use `model` as the forward function. 

Should we update according to be consistent with the recommendation?  What do you think?

[1]: https://github.com/ankurtaly/Integrated-Gradients/blob/master/howto.md#identifying-the-output-tensor
[2]: https://captum.ai/tutorials/IMDB_TorchText_Interpret

**Comment by NarineK** [comment]
Hi @heytitle, this is very interesting! Thank you for bringing it up.
IMDB classification task is a binary classification, not multi-class (assuming the definition of multi-class when the number of classes are larger than 2 ) 

We can definitely also try the attribution for sigmoid space, but we will get similar results. At least I'm seeing similar results.

**Comment by wangyongjie-ntu** [comment]
@heytitle I have the same concerns. Why do these methods in Captum use the logits instead of the probability after softmax?  
In my view,  these methods hold on arbitrary functions(before or after softmax). Is there some insight behind it?

**Comment by NarineK** [comment]
@wangyongjie-ntu , that is an excellent question. Usually when we compute the attribution using the softmax or sigmoid layers as an output layer, we obtain smaller attribution scores. 

Specifically, the author's of deeplift paper recommend to compute the attributions with respect to the logit layer:
https://arxiv.org/pdf/1704.02685.pdf
Section 3.6. Choice of Target Layer discusses that case.
Here the authors make the case that if we use the sigmoid, higher logit doesn't translate into a higher attribution score.

In case of IG, in the tutorials above I've noticed that those scores are comparable.

**Comment by wangyongjie-ntu** [comment]
@NarineK . Smaller attribution is right because the difference of probability is less than it of logits (the completeness holds).  
However, when the captum deeplift uses the probability as the function F, the completeness seems to be violated. Here is an example when analyzing the feature importance on a naive Multi-layer perception.  With logits, it is OK.

![image](https://user-images.githubusercontent.com/12019443/85940788-2efc5600-b951-11ea-90a3-40887bac7296.png)

**Comment by NarineK** [comment]
@wangyongjie-ntu , is the last layer a sigmoid or softmax ?

**Comment by NarineK** [comment]
Hi @wangyongjie-ntu, we apply rules on both softmax and sigmoid and we are able to apply the rules only if they are defined through torch.nn in the init of the module. `e.g. self.softmax = torch.nn.Softmax()`
If we define them through Functionals we will not be able to hook them.

**Comment by wangyongjie-ntu** [comment]
@NarineK  Do you mean this kind of code?

`class Mobile(nn.Module):

    def __init__(self):
        super(Mobile, self).__init__()
        self.model = nn.Sequential(
                nn.Linear(20, 16),
                nn.ReLU(),
                nn.Linear(16, 12),
                nn.ReLU(),
                nn.Linear(12, 4)
                )
        self.softmax = nn.Softmax(dim = 1)

    def forward(self, x):

        fc = self.model(x)
        max_factor, _ = torch.max(fc, dim = 1)
        max_factor = max_factor.expand(4, len(max_factor)).t()
        normed_fc = fc - max_factor
        prob = self.softmax(normed_fc)
        prob = prob + 10e-10

        return prob
`

**Comment by NarineK** [comment]
@wangyongjie-ntu, yes, that's what I meant by defining `self.softmax` in the constructor. I see that you take also, torch.max. Deeplift implementation overrides the gradients for non-linearities but not for torch.max. I'm a bit concerned about the `torch.max ` because it won't be overidden, we only override Maxpools.

**Comment by wangyongjie-ntu** [comment]
@NarineK  I use the probability in the above class, but deepLIFT still computes the feature important w.r.t. the logits.

**Comment by NarineK** [comment]
@wangyongjie-ntu, how do you define DeepLift constructor in that case ? The original question was that it is working with logits but doesn't with softmax ?

**Comment by wangyongjie-ntu** [comment]
@NarineK Thanks very much for your patient reply. 

The image in my second comment is the model built with torch.nn.functional.softmax(). The deepLIFT returns the smaller importance score violating the completeness. As you mentioned, hook function only works on the torch.nn module, Hence, I updated the code as above snippets. It seems that deepLIFT still views the logits as the function rather than the probability defined in the snippets, even if I remove the torch.max operation and directly feed the fc into nn.softmax(). 

Why do I want to use DeepLIFT on prob? 
I proposed a draft method on **probability** and selected the IG, deepLIFT etc. as baselines.  To fairly compare the performance, I want to compute the importance score on the probability function.

**Comment by NarineK** [comment]
Hi @wangyongjie-ntu, thank you for the explanation!  Nice, directly feeding the fc into torch.nn.softmax sounds like a good idea. Glad that DeepLift worked.

**Comment by NarineK** [comment]
@wangyongjie-ntu, can we close this issue or are you still working on it ?

**Comment by wangyongjie-ntu** [comment]
@NarineK Thanks. We can close.
